docs: add organizations, provisioners, and premium license docs (#14778)
coder/coder: docs/admin/auth.md

# Authentication

![OIDC with Coder Sequence Diagram](../images/oidc-sequence-diagram.svg).

> 

Reviewer: Do we have an issue to create a new diagram?

Author: No I just want to remove it for now

![Unauthorized group error](../images/admin/group-allowlist.png)

## Role sync (enterprise) (premium)

If your OpenID Connect provider supports roles claims, you can configure Coder
to synchronize roles in your auth provider to roles within Coder.

> There are two ways you can configure group sync:

Reviewer 2: ```suggestion
There are two ways you can configure role sync:
```

> There are two ways you can configure group sync:

Reviewer 2: There are 2 ways to do role sync. Server Flags assign **site wide** roles, and runtime org role sync assigns **organization roles**

> One role from your identity provider can be mapped to many roles in Coder
> (e.g. the example above maps to 2 roles in Coder.)

## Runtime (Organizations)

> Note: You must have a Premium license with Organizations enabled to use this.
> [Contact your account team](https://coder.com/contact) for more details

For deployments with multiple [organizations](./organizations.md), you can
configure role sync at the organization level. In future Coder versions, you
will be able to configure this in the UI. For now, you must use CLI commands.

First, confirm that your OIDC provider is sending a roles claim by logging in
with OIDC and visiting the following URL:

```text
https://[coder.example.com]/api/v2/debug/[your-username]/debug-link

> ```

Reviewer 2: This will only work for `owner`s

You should see a field in either `id_token_claims`, `user_info_claims` or both
followed by a list of the user's OIDC roles in the response. This is the
[claim](https://openid.net/specs/openid-connect-core-1_0.html#Claims) sent by
the OIDC provider. See
[Troubleshooting](#troubleshooting-grouproleorganization-sync) to debug this.

> Depending on the OIDC provider, this claim may be named differently.

To fetch the current group sync settings for an organization, run the following:

```sh
coder organizations settings show role-sync \
  --org <org-name> \
  > role-sync.json
```

The default for an organization looks like this:

```json
{
	"field": "",
	"mapping": null
}
```

Below is an example that uses the `roles` claim and maps `coder-admins` from the
IDP as an `Organization Admin` and also maps to a custom `provisioner-admin`
role.

```json
{
	"field": "roles",
	"mapping": {
		"coder-admins": ["organization-admin"],
		"infra-admins": ["provisioner-admin"]
	}
}
```

> Note: Be sure to use the `name` field for each role, not the display name. Use
> `organization-admin`, `organization-user-admin`, `organization-auditor` for

> > the built in roles.

Reviewer 2: `coder organization  roles show --org=coder` can list all roles in a org

To set these role sync settings, use the following command:

```sh
coder organizations settings set role-sync \
  --org <org-name> \
  < role-sync.json
```

Visit the Coder UI to confirm these changes:

![IDP Sync](../images/admin/organizations/role-sync.png)

</div>

> ## Organization Sync (Premium)

Reviewer 2: Leave a note the experience of editing this will be improved. Likely a UI and changable at runtime in the future.

If your OpenID Connect provider supports groups/role claims, you can configure

> Coder to synchronize roles in your auth provider to organizations within Coder.

Reviewer 2: ```suggestion
Coder to synchronize claims in your auth provider to organizations within Coder.
```

> First, confirm that your OIDC provider is sending a groups claim by logging in

Reviewer 2: ```suggestion
First, confirm that your OIDC provider is sending claims by logging in
```

First, confirm that your OIDC provider is sending a roles claim by logging in
with OIDC and visiting the following URL:

```text
https://[coder.example.com]/api/v2/debug/[your-username]/debug-link
```

You should see a field in either `id_token_claims`, `user_info_claims` or both
followed by a list of the user's OIDC roles in the response. This is the
[claim](https://openid.net/specs/openid-connect-core-1_0.html#Claims) sent by
the OIDC provider. See [Troubleshooting](#troubleshooting-grouprole-sync) to
debug this.

> Depending on the OIDC provider, this claim may be named differently.

To fetch the current group sync settings for an organization, run the following:

```sh
coder organizations settings show role-sync \
  --org <org-name> \
  > role-sync.json
```

The default for an organization looks like this:

```json
{
	"field": "",
	"mapping": null
}
```

Below is an example that uses the `roles` claim and maps `coder-admins` from the
IDP as an `Organization Admin` and also maps to a custom `provisioner-admin`
role:

```json
{
	"field": "roles",
	"mapping": {
		"coder-admins": ["organization-admin"],
		"infra-admins": ["provisioner-admin"]

> 	}

Author: @Emyrk - For custom roles, do you use the `name` field? I noticed there wasn't really validation for role sync

Author: edit: nvm I confirmed it
